source/extensions/quic/connection_id_generator/quic_lb/quic_lb.cc

```diff
@@ -0,0 +1,353 @@
+    RETURN_IF_NOT_OK_REF(result);
+
+    // Record the length of the encoder's connection ID so that we know the offset to look for
+    // routing information.
+    ret->encoder_connection_id_length_ =
+        test_instance_or_result.value()->encoder_.GenerateConnectionId().length();
+  }
+
+  ret->secrets_provider_ =
+      secretsProvider(config.encryption_parmeters(), context.getTransportSocketFactoryContext(),
+                      context.initManager());
+  if (ret->secrets_provider_ == nullptr) {
+    return absl::InvalidArgumentError("invalid encryption_parmeters config");
+  }
+
+  ret->secrets_provider_validation_callback_handle_ = ret->secrets_provider_->addValidationCallback(
+      [&api = context.serverFactoryContext().api()](
+          const envoy::extensions::transport_sockets::tls::v3::GenericSecret& secret)
+          -> absl::Status { return getAndValidateKeyAndVersion(secret, api).status(); });
+
+  ret->secrets_provider_update_callback_handle_ = ret->secrets_provider_->addUpdateCallback(
+      [&factory = *ret, &api = context.serverFactoryContext().api()]() -> absl::Status {
+        const envoy::extensions::transport_sockets::tls::v3::GenericSecret* secret =
+            factory.secrets_provider_->secret();
+        if (secret == nullptr) {
+          return absl::NotFoundError("secret update callback called with empty secret");
+        }
+
+        auto data_or_result = getAndValidateKeyAndVersion(*secret, api);
+        RETURN_IF_NOT_OK_REF(data_or_result.status());
+
+        factory.tls_slot_->runOnAllThreads(
+            [data =
+                 data_or_result.value()](OptRef<QuicLbConnectionIdGenerator::ThreadLocalData> obj) {
+              ASSERT(obj.has_value(),
+                     "Guaranteed if `set()` was previously called on the tls slot");
+
+              // TODO: error handling from both of these calls
+              auto result = obj->updateKeyAndVersion(data.first, data.second);
+              ASSERT(result.ok(), "Parameters and data was already validated");
+            });
+
+        return absl::OkStatus();
+      });
+
+  ret->tls_slot_ = ThreadLocal::TypedSlot<QuicLbConnectionIdGenerator::ThreadLocalData>::makeUnique(
+      context.serverFactoryContext().threadLocal());
+
+  // using InitializeCb = std::function<std::shared_ptr<T>(Event::Dispatcher & dispatcher)>;
+  ret->tls_slot_->set(
+      [=](Event::Dispatcher&) -> std::shared_ptr<QuicLbConnectionIdGenerator::ThreadLocalData> {
+        auto result = QuicLbConnectionIdGenerator::ThreadLocalData::create(config, server_id);
+        ASSERT(result.status().ok()); // Configuration was validated above.
+        return result.value();
+      });
+
+  return ret;
+}
+
+QuicConnectionIdGeneratorPtr Factory::createQuicConnectionIdGenerator(uint32_t) {
+  return std::make_unique<QuicLbConnectionIdGenerator>(*tls_slot_);
+}
+
+Network::Socket::OptionConstSharedPtr
+Factory::createCompatibleLinuxBpfSocketOption(uint32_t concurrency) {
```

Comment:
Is this still WIP? The filter in this function seems to be identical to `EnvoyDeterministicConnectionIdGeneratorFactory::createCompatibleLinuxBpfSocketOption`, which (I think) doesn't understand QUIC-LB CIDs.

Reply:
Yes this is still WIP, and you're correct that it is a copy/paste of the other one, and this will not work at all for these CIDs.

Reply:
@wu-bin this is now getting closer to ready (but still needs more test coverage). For the initial version to unblock interoperability testing, I'm going to entirely omit the BPF filter and accept poor performance, and then add the BPF filter a bit later.